app/models/taxon_change.rb

```diff
@@ -127,6 +131,10 @@ def commit
     unless committable_by?( committer )
       raise PermissionError, "Committing user doesn't have permission to commit"
     end
+    if rank_level_conflict?
+      Rails.logger.error "[ERROR #{Time.now}] Failed to commit records for #{self}: all input taxon descendant rank levels must be finer than the output taxon rank level"
+      return
```

Comment:
should I make a new custom type of error like 'RankLevelError' or is there a general error I can/should co-opt?

Reply:
RankLevelError will provide some very specific information in our crash logs, so that would be good, but what I really want is some feedback for the user, so whatever you do, the controller action that commits the record needs to be aware when this problem happens, e.g. it needs to be ready to rescue from this exception, or it needs to perform the rank level check before committing, or rank level needs to be tied to the TaxonChange's validation and the controller should re-check validity before committing, or something. For checking permission we get around this by just not showing the "Commit" button if the current user can't commit the change, but that may not be appropriate here.

IMO, what I would do is make a new RankLevelError subclass of StandardError (see the way PermissionError is defined earlier in the file) and raise that if there's a rank problem when you try to commit. That will ensure that no matter how the commit gets triggered, it will not proceed if this problem exists. I would also do something in the controller to check for input descendants with rank problems and return an error message that names some of the problem descendants before even trying to commit.